nasa/ow_autonomy
# Digging procedure, power monitoring, plan refactoring, internals

**Features**

1.  Digging Procedure (OCEANWATER-429, OCEANWATER-427), as documented [here](https://babelfish.arc.nasa.gov/confluence/pages/viewpage.action?pageId=103941080).

2. A simplified reference mission plan (ReferenceMission1) that uses the digging procedure above.

3. Power monitoring (no Jira, though prerequisite for OCEANWATER-47).  Periodic logs of state of charge and remaining useful battery life.

**Internal improvements**

1. Use of Plexil contruct SynchronousCommand.
   This greatly simplifies the synchronous execution of any lander operation.  Reduced the size and complexity of many Plexil plans (.plp files), but added a fair amount of C++ support.

2.  Much plan refactoring and removal of most "demo" plans.

3.  Incorporation of new Plexil adapter configuration for commands, replacing OwInterface::executeCommand function.

**Dependencies**

1. Requires two branches of ow_simulator (these were merged together for testing):
 - OCEANWATER-475 for power (RUL)
 - OCEANWATER-476 for Grind enhancements

2. Requires OCEANWATER-476 branch of ow_autonomy (one version already merged in).

**Testing**

1. Start simulator, e.g. europa_terminator_workspace.launch

2. Start autonomy node: `roslaunch ow_autonomy autonomy_node.launch plan:=ReferenceMission1.plx`

The lander should:
1. Perform a brief (truncated) imaging of landing site.
2. Use a Guarded Move to find ground (location arbitrary, may be broken - see Caveats below).
3. Perform the digging procedure as described [here](https://babelfish.arc.nasa.gov/confluence/pages/viewpage.action?pageId=103941080), unless step 2 encounters error finding ground.

**Caveats**

1. GuardedMove is often not finding ground properly, yet may be returning success anyway.  This is breaking plans ReferenceMission1 and TorqueTest.

**Code Review**

A few files have changed so much that it may be easier to review the file as a whole, instead of the diffs.

Comments from earlier code reviews of ow_autonomy have not all been addressed, e.g. there are still raw pointers in the classes (and a separate Jira to address this).

There are several instances of repetitive code patterns for which I do not know a more compact solution offhand.  Any suggestions appreciated.
